### ISSUE TITLE
Use data kind for optic flavours?

### ISSUE THREAD

**Issue by adamgundry** [well-typed/optics#24]
At the moment we use empty types as optic flavour tags (`A_Lens` etc.), but with the reorganisation in #19 we can't extend the hierarchy later, so we don't need an open kind. (Such extensions were never really well-supported anyway, because of `Join`.) Would it be worth switching to a promoted data type? That would rule out ill-formed types such as `Optic Int i o s t a b`.

**Comment by phadej** [comment]
@arybczak not necessarily.

```haskell
-- if we don't pollute both type and term namespaces with same name...
data FooBar = Foo | Bar

-- we can make aliases
type Foo = 'Foo
type Bar = 'Bar
```

:+1: from me to switch to data-kind.

**Comment by phadej** [comment]
Implemented in #36/#35

**Comment by phadej** [comment]
And reverted, as we didn't like ticked versions (in error messages)
